// dartjs/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.injectSmoothVolume = exports.VolumeTransformer = exports.VoiceReceiver = exports.VoiceConnection = exports.StreamDispatcher = exports.DartVoiceManager = void 0;
const injection_1 = require("./smoothVolume/injection");
Object.defineProperty(exports, "injectSmoothVolume", { enumerable: true, get: function () { return injection_1.injectSmoothVolume; } });
if (!("DARTJS_DISABLE_INJECTION" in process.env)) {
    (0, injection_1.injectSmoothVolume)();
}
var DartVoiceManager_1 = require("./core/DartVoiceManager");
Object.defineProperty(exports, "DartVoiceManager", { enumerable: true, get: function () { return DartVoiceManager_1.DartVoiceManager; } });
var StreamDispatcher_1 = require("./core/StreamDispatcher");
Object.defineProperty(exports, "StreamDispatcher", { enumerable: true, get: function () { return StreamDispatcher_1.StreamDispatcher; } });
var VoiceConnection_1 = require("./core/VoiceConnection");
Object.defineProperty(exports, "VoiceConnection", { enumerable: true, get: function () { return VoiceConnection_1.VoiceConnection; } });
var VoiceReceiver_1 = require("./core/VoiceReceiver");
Object.defineProperty(exports, "VoiceReceiver", { enumerable: true, get: function () { return VoiceReceiver_1.VoiceReceiver; } });
var VolumeTransformer_1 = require("./smoothVolume/VolumeTransformer");
Object.defineProperty(exports, "VolumeTransformer", { enumerable: true, get: function () { return VolumeTransformer_1.VolumeTransformer; } });
__exportStar(require("./types/types"), exports);
__exportStar(require("./Utils/Util"), exports);
